doc/design/polymorphic_handler.md

```diff
@@ -0,0 +1,243 @@
+- concurrent access of the instance is delayed until the initializing call completes instance
+  initialization
+- instance can be accessed like a regular singleton
+
+The construction also creates a `StaticLifetimeGuard` with static lifetime, the primary guard. This ensures basic
+static lifetime of the constructed instance (as if it would have been a static variable in a Meyers
+singleton itself).
+
+### Reference counting
+
+- global static atomic reference counter (initially zero)
+- construction (including copy/move) of each guard object increases the counter by one
+- destruction decreases the counter by one
+- if the counter reaches zero (last guard destroyed), the instance is  destroyed **IF it was
+  constructed before**
+
+### Instance destruction
+
+The instance is only constructed when it is needed (i.e. lazily). There can be several
+guard objects without an instance ever being constructed. In this case no instance destruction takes
+place once the counter reaches zero.
+
+Due to atomic counter decrement, the instance is destroyed exactly once.
+
+### Instance construction after destruction
+
+It is not possible to replace the instance once constructed due to the static primary guard.
+Technically it would be possible by some static destructor after the primary guard (and all other
+guards) are destroyed, but this happens only during program termination.
+
+## Using the Polymorphic Handler
+
+### Basic Usage
+
+```cpp
+struct Interface {
+public:
+    virtual void foo() = 0;
+};
+
+struct DefaultHandler : public Interface {
+    void foo() override { /* ... */ }
+};
+
+struct OtherHandler : public Interface {
+    void foo() override { /* ... */ }
+};
+
+using Handler = PolymorphicHandler<Interface, DefaultHandler>;
+```
+
+The first time we access the handler, it is initialized as a DefaultHandler that is guarded 
+internally with a `StaticLifeTimeGuard`.
+
+```cpp
+// thread 1
+// get a reference to the current handler
+auto& handler = Handler::get();
+
+// use the handler polymorphically
+handler.foo();
+```
+
+A concurrent thread 2 may switch the handler to some other handler. This will not interfere with the 
+execution of `foo` in thread 1, which is still using the old handler (which is ensured to be alive
+by a guard).
+
+The default handler is part of the `PolymorphicHandler`, but any other handler to be set is not. To
+ensure the lifetime of handlers that are used, the API to set another handler requires using a
+guard. The guard is passed by value and an internal copy ensures any handler being set is not
+destroyed before main exits. Any further lifetime must be controlled with external guards.
+
+```cpp
+// thread 2
+StaticLifetimeGuard<OtherHandler> guard;
+
+// the OtherHandler instance may not be constructed yet
+
+// set the handler to the instance guarded by guard,
+// this will create another guard to ensure the lifetime
+Handler::set(guard);
```

Comment:
Consider to check the return value and give a hint that `set` would fail after finalize.

Reply:
Added code that suggests that in an explicit check.